consensus/src/persistent_liveness_storage.rs

```diff
@@ -90,21 +96,120 @@ impl LedgerRecoveryData {
+        blocks.sort_by_key(|b| (b.epoch(), b.round()));
+
+        let latest_commit_idx = blocks
+            .iter()
+            .position(|block| block.id() == latest_commit_id)
+            .ok_or_else(|| format_err!("unable to find root: {}", latest_commit_id))?;
+        let commit_block = blocks[latest_commit_idx].clone();
+        let root_quorum_cert = quorum_certs
+            .iter()
+            .find(|qc| qc.certified_block().id() == commit_block.id())
+            .ok_or_else(|| format_err!("No QC found for root: {}", commit_block.id()))?
+            .clone();
+
+        let (root_ordered_cert, root_commit_cert) = if order_vote_enabled {
+            // We are setting ordered_root same as commit_root. As every committed block is also ordered, this is fine.
+            // As the block store inserts all the fetched blocks and quorum certs and execute the blocks, the block store
+            // updates highest_ordered_cert accordingly.
+            let root_ordered_cert =
+                WrappedLedgerInfo::new(VoteData::dummy(), latest_ledger_info_sig.clone());
+            (root_ordered_cert.clone(), root_ordered_cert)
+        } else {
+            let root_ordered_cert = quorum_certs
+                .iter()
+                .find(|qc| qc.commit_info().id() == commit_block.id())
+                .ok_or_else(|| format_err!("No LI found for root: {}", latest_commit_id))?
+                .clone()
+                .into_wrapped_ledger_info();
+            let root_commit_cert = root_ordered_cert
+                .create_merged_with_executed_state(latest_ledger_info_sig)
+                .expect("Inconsistent commit proof and evaluation decision, cannot commit block");
+            (root_ordered_cert, root_commit_cert)
+        };
+
+        let window_start_round = blocks[latest_commit_idx]
+            .round()
+            .saturating_add(1)
+            .saturating_sub(window_size);
+        let mut id_to_blocks = HashMap::new();
+        blocks.iter().for_each(|block| {
+            id_to_blocks.insert(block.id(), block);
+        });
+
+        let mut curr_id = latest_commit_id;
+        let mut window_start_id = HashValue::zero();
+        while let Some(block) = id_to_blocks.get(&curr_id) {
+            if block.quorum_cert().certified_block().round() < window_start_round {
+                window_start_id = curr_id;
+                break;
+            }
+
+            window_start_id = curr_id;
+            curr_id = block.parent_id();
+        }
+        ensure!(
+            window_start_id != HashValue::zero(),
+            "Window start block not found"
+        );
+
+        let window_start_idx = blocks
+            .iter()
+            .position(|block| block.id() == window_start_id)
+            .ok_or_else(|| format_err!("unable to find root: {}", window_start_id))?;
+        let window_start_block = blocks.remove(window_start_idx);
```

Comment:
why do we remove the window_start_block but not commit block, seems inconsistent

Reply:
still a question, because there's a redundancy check when insert committed block?